File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Rwanda.m
//
//  UIImage+Rwanda.m
//  CountriesFlags
//

#import "Countries/UIImage+Rwanda.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (Rwanda)

+ (UIImage *)flagRwandaWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:163.0f/255.0f
                                          blue:224.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:250.0f/255.0f
                                           green:210.0f/255.0f
                                            blue:1.0f/255.0f
                                           alpha:1.0f];
    UIColor* colorGreen = [UIColor colorWithRed:32.0f/255.0f
                                          green:96.0f/255.0f
                                           blue:61.0f/255.0f
                                          alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 130.0f * 97.5f, size.width, size.height / 130.0f * 32.5f));

    [colorYellow setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 130.0f * 65.0f, size.width, size.height / 130.0f * 32.5f));
    CGPoint starCenter = CGPointMake(size.width / 195.0f * 160.0f, size.height / 130.0f * 34.0f);
    [[UIBezierPath starAtPoint:starCenter
                       radius:size.height / 130.0f * 42.0f / 2.0f
                  innerRadius:size.height / 130.0f * 13.0f / 2.0f
                         type:FlagStarTypeTwentyFourPointed] fill];

    [colorBlue setStroke];
    UIBezierPath* path = [UIBezierPath bezierPathWithArcCenter:starCenter
                                                        radius:size.height / 130.0f * 12.2f / 2.0f
                                                    startAngle:0.0f
                                                      endAngle:M_PI * 2.0f
                                                     clockwise:YES];
    path.lineWidth = size.height / 145.0f;
    [path stroke];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Rwanda.swift
//
//  UIImage+Rwanda.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Rwanda "wikipedia"
     
     Generates the flag of Rwanda.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Rwanda.
     */
    internal static func flagRwanda(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 0.0,
                                green: 163.0 / 255.0,
                                blue: 224.0 / 255.0,
                                alpha: 1.0)
        let colorYellow = UIColor(red: 250.0 / 255.0,
                                  green: 210.0 / 255.0,
                                  blue: 1.0 / 255.0,
                                  alpha: 1.0)
        let colorGreen = UIColor(red: 32.0 / 255.0,
                                 green: 96.0 / 255.0,
                                 blue: 61.0 / 255.0,
                                 alpha: 1.0)
        let lineSize = CGSize(width: size.width, height: size.height / 130.0 * 32.5)

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorGreen.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 130.0 * 97.5), size: lineSize))

        colorYellow.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 130.0 * 65.0), size: lineSize))

        let starCenter = CGPoint(x: size.width / 195.0 * 160.0, y: size.height / 130.0 * 34.0)
        UIBezierPath.star(at: starCenter,
                          radius: size.height / 130.0 * 42.0 / 2.0,
                          type: .twentyFourPointed(innerRadius: size.height / 130.0 * 13.0 / 2.0)).fill()

        colorBlue.setStroke()
        let path = UIBezierPath(arcCenter: starCenter,
                                radius: size.height / 130.0 * 12.2 / 2.0,
                                startAngle: 0.0,
                                endAngle: .pi * 2.0,
                                clockwise: true)
        path.lineWidth = size.height / 145.0
        path.stroke()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
